docs/strings.md

```diff
@@ -0,0 +1,529 @@
+The current version of the Unicode specification, 13.0, uses just over 10% of this encoding space and even with the amount wasted on emoji it seems unlikely that the total space will be exhausted and so Verona is likely to be in a in a fortunate position of being able to define a string API without the definition of a character changing substantially over the lifetime of the language.
+
+It is incredibly unlikely; however, that a Verona string implementation will contain native Verona code that implements the full Unicode specification in an early release.
+The specification of the [Unicode collation algorithm](https://www.unicode.org/reports/tr10/) alone runs to several hundred pages of text.
+A string design will therefore need to be efficient when interoperating with a sandboxed C++ library, such as ICU.
+
+Who cares about strings?
+------------------------
+
+Verona aims to be an *infrastructure programming language*.
+This category encompasses target use cases such as device drivers, distributed databases, online translation engines, and so on.
+Some of these will not have any user-facing component and so will use strings as identifiers (for example, dictionary keys, path components, error indicators) without the need for any localisation.
+Others will use strings to present text to a human and will need rich unicode processing.
+Our goal should be to support both use cases with the same set of abstractions but avoid projects in the former category needing to pay any performance or code-size penalty relative to a simple array-of-probably-ASCII-bytes string representation.
+
+A lot of complex Verona systems are likely to include components that are at both extremes in terms of their use of strings.
+We should not require a global decision about how strings are represented, nor should we penalise modules that require only simple strings.
+
+Code units, code points, characters, and grapheme clusters
+----------------------------------------------------------
+
+Strings are, informally, sequences of characters.
+This is not always a useful abstraction as the definition of a character can, in different contexts, mean:
+
+ * A code unit, the smallest indivisible unit of a single or multi-byte encoding (e.g. `char` or `wchar_t` in C, `unichar` in Objective-C)
+ * A Unicode code point, which may be stored as 1-4 octets, depending on the encoding.
+ * A [grapheme cluster](https://www.unicode.org/reports/tr29/), a sequence of code points that are combined to display something that a human will consider to be a character.
+
+Iteration over strings is therefore very complex.
+Should the unit of iteration be a code unit, a code point, or a grapheme cluster?
+Should iteration canonicalise any of these?
+For example, é can be represented as either a single code point or as an e followed by with an accent combining diacritic, but both representations are regarded as equivalent by unicode.
```

Comment:
Just thinking out loud, I think we should canonicalise everything we possibly can, using the standard's own rules or best practices, and documenting when it's not clear or when we pick a specific default. This may result in similar strings being iterated or accesses in different ways for different standards, but in the vast majority of cases people use only one standard and convert everything else into it.

Reply:
Unfortunately, it's not always clear that this is a good idea.  If all you want to do is stream processing on a string, canonicalization is unnecessary and expensive.  One of the reasons strings are so annoying is that simple decisions like this can have huge performance implications on the macro scale, both positive and negative, depending on the workload.